agent/xds/builtinextensions/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,167 @@
+	envoy_type_v3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	envoy_resource_v3 "github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"github.com/hashicorp/go-multierror"
+	"github.com/mitchellh/mapstructure"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	"github.com/hashicorp/consul/agent/xds/builtinextensiontemplate"
+	"github.com/hashicorp/consul/agent/xds/xdscommon"
+	"github.com/hashicorp/consul/api"
+)
+
+type ratelimit struct {
+	ProxyType     string
+	Listener      string
+	MaxTokens     int
+	TokensPerFill int
+	FillInterval  int
+}
+
+var _ builtinextensiontemplate.Plugin = (*ratelimit)(nil)
+
+// MakeRatelimitExtension is a builtinextensiontemplate.PluginConstructor for a builtinextensiontemplate.EnvoyExtension.
+func MakeRatelimitExtension(ext xdscommon.ExtensionConfiguration) (builtinextensiontemplate.Plugin, error) {
+	var resultErr error
+	var plugin ratelimit
+
+	if name := ext.EnvoyExtension.Name; name != api.BuiltinRatelimitExtension {
+		return nil, fmt.Errorf("expected extension name 'ratelimit' but got %q", name)
+	}
+
+	if err := mapstructure.Decode(ext.EnvoyExtension.Arguments, &plugin); err != nil {
+		return nil, fmt.Errorf("error decoding extension arguments: %v", err)
+	}
+
+	if plugin.FillInterval == 0 {
+		resultErr = multierror.Append(resultErr, fmt.Errorf("fillInterval is required"))
+	}
+
+	if plugin.MaxTokens == 0 {
+		resultErr = multierror.Append(resultErr, fmt.Errorf("MaxTokens must be greater than 0"))
+	}
+
+	if plugin.TokensPerFill == 0 {
+		resultErr = multierror.Append(resultErr, fmt.Errorf("TokensPerFill must be greater than 0"))
+	}
+
+	if err := validateProxyType(plugin.ProxyType); err != nil {
+		resultErr = multierror.Append(resultErr, err)
+	}
+
+	if err := validateListener(plugin.Listener); err != nil {
+		resultErr = multierror.Append(resultErr, err)
+	}
+
+	return plugin, resultErr
+}
+
+// CanApply determines if the extension can apply to the given extension configuration.
+func (p ratelimit) CanApply(config xdscommon.ExtensionConfiguration) bool {
+	return string(config.Kind) == p.ProxyType && p.matchesListenerDirection(config)
+}
+
+func (p ratelimit) matchesListenerDirection(config xdscommon.ExtensionConfiguration) bool {
+	return !config.IsUpstream() && p.Listener == "inbound"
+}
+
+// PatchRoute does nothing.
+func (p ratelimit) PatchRoute(route *envoy_route_v3.RouteConfiguration) (*envoy_route_v3.RouteConfiguration, bool, error) {
+	return route, false, nil
+}
+
+// PatchCluster does nothing.
+func (p ratelimit) PatchCluster(c *envoy_cluster_v3.Cluster) (*envoy_cluster_v3.Cluster, bool, error) {
+	return c, false, nil
+}
+
+// PatchFilter inserts a http local rate_limit filter at the head of
+// envoy.filters.network.http_connection_manager filters
+func (p ratelimit) PatchFilter(filter *envoy_listener_v3.Filter) (*envoy_listener_v3.Filter, bool, error) {
```

Comment:
General question (maybe more directed for @erichaberkorn). Is it the programmer's responsibility to order extensions so they insert themselves in the right place in the chain?

What happens if two filters conflict when trying to be first in a chain?

Reply:
We handle these issues in a couple of ways:
* Extensions should error out when patching resources that don't look as expected.
* We will add support for post validations to extensions. This would handle cases where an extension applies correctly and a later extension messes up the Envoy resources.